client/asset/eth/eth.go

```diff
@@ -1463,13 +1435,20 @@ func (w *ETHWallet) FundOrder(ord *asset.Order) (asset.Coins, []dex.Bytes, error
 
 // FundOrder locks value for use in an order.
 func (w *TokenWallet) FundOrder(ord *asset.Order) (asset.Coins, []dex.Bytes, error) {
-
 	if w.gasFeeLimit() < ord.MaxFeeRate {
 		return nil, nil, fmt.Errorf(
 			"%v: server's max fee rate %v higher than configured fee rate limit %v",
 			dex.BipIDSymbol(w.assetID), ord.MaxFeeRate, w.gasFeeLimit())
 	}
 
+	approvalStatus, err := w.approvalStatus(ord.Version)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error getting approval status: %v", err)
+	}
+	if approvalStatus != asset.Approved {
+		return nil, nil, fmt.Errorf("token %d not approved for trading", w.assetID)
```

Comment:
I agree that a new error type is warranted. If clients are already doing special error handling based on the new error type, then might as well have them also do an `ApproveToken` call I think. I'm not totally against it, but it feels kind of weird that `FundOrder` spends money without it having been approved by the caller. The split transaction spends money, but that was at least approved by the wallet settings or the order config.

I'm thinking it would be helpful to add an `onConfirm func()` argument to the `ApproveToken` function, then clients can easily know when it's ready to trade.

Reply:
Good points.  Clients will just have to adapt to this change and call the approve functions themselves.